### PR TITLE
gitlab_project: wait until real deletion

### PR DESCRIPTION
**Pull request by roidelapluie**
Deleting a project in Gitlab is an async method.

By blocking until the project does not exist anymore, we
do not lie to the terraform end user about the status of a
deleted project.

As a side effet it fixes the acceptances tests.

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
# Before

```
$ make testacc                                                          
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (2.45s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (1.01s)
=== RUN   TestAccGitlabProjectHook_basic
--- FAIL: TestAccGitlabProjectHook_basic (1.27s)
        testing.go:499: Error destroying resource! WARNING: Dangling resources
                may exist. The full state and error is shown below.

                Error: Check failed: Repository still exists

                State: <no state>
=== RUN   TestAccGitlabProject_basic
--- FAIL: TestAccGitlabProject_basic (1.06s)
        testing.go:499: Error destroying resource! WARNING: Dangling resources
                may exist. The full state and error is shown below.

                Error: Check failed: Repository still exists

                State: <no state>
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-gitlab/gitlab 5.793s
make: *** [GNUmakefile:15: testacc] Error 1

```

**Comment by roidelapluie** [comment]
# After

```
$ make testacc 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (4.78s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (0.90s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (2.08s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (2.22s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 9.989s
```

File: gitlab/resource_gitlab_project.go [hunk 1]
```diff
@@ -196,5 +197,16 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
...
-	return err
+	if err != nil {
+		return err
+	}
+	for {
+		_, response, err := client.Projects.GetProject(d.Id())
+		if err != nil {
+			if response.StatusCode == 404 {
+				return nil
+			}
```
Comment (tombuildsstuff) [inline gitlab/resource_gitlab_project.go]: do we not want to return this error if it's not a 404?

File: gitlab/resource_gitlab_project.go [hunk 2]
```diff
@@ -196,5 +197,16 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
...
+		}
+		time.Sleep(time.Second * 1)
```
Comment (tombuildsstuff) [inline gitlab/resource_gitlab_project.go]: it'd be good to add a max attempts timeout here, if possible? i.e. try 30 times then raise an error?

Comment (roidelapluie) [reply gitlab/resource_gitlab_project.go]: I was relying on terraform to do this, e.g. time out?

Comment (roidelapluie) [reply gitlab/resource_gitlab_project.go]: isnt that a terraform core functionnality?

Comment (tombuildsstuff) [reply gitlab/resource_gitlab_project.go]: Terraform exposes this as the `context` but I thought it needed to be threaded through? (I may be wrong!)

In general we'd [tend to achieve this using `StateChangeConf`](https://github.com/terraform-providers/terraform-provider-vsphere/blob/master/vsphere/resource_vsphere_datacenter.go#L64) to ensure it's deleted

**Comment by roidelapluie** [comment]
@tombuildsstuff is this better?

**Comment by roidelapluie** [comment]
```
$ make testacc 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (5.88s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (0.60s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (5.83s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (5.84s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 18.151s
```

File: gitlab/resource_gitlab_project.go (hunk 2 above)
Comment (roidelapluie) [reply gitlab/resource_gitlab_project.go]: I have implemented that. Waiting for your review to merge.

File: gitlab/resource_gitlab_project.go [hunk 3]
```diff
@@ -196,5 +198,35 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
 	log.Printf("[DEBUG] Delete gitlab project %s", d.Id())
 
 	_, err := client.Projects.DeleteProject(d.Id())
-	return err
+	if err != nil {
+		return err
+	}
+
+	// Wait for the datacenter resource to be ready
```
Comment (richardc) [inline gitlab/resource_gitlab_project.go]: Datacenter?  I suspect this is dodgy copy and paste.

**Comment by roidelapluie** [comment]
Thanks :)

**Pull request merged**